src/mame/casio/cz230s.cpp

```diff
@@ -0,0 +1,693 @@
+
+	save_item(NAME(m_port_a));
+	save_item(NAME(m_key_sel));
+	save_item(NAME(m_key_mux));
+	save_item(NAME(m_midi_rx));
+}
+
+/**************************************************************************/
+void cz230s_state::machine_reset()
+{
+	m_key_sel = m_key_mux = 0;
+	m_midi_rx = 1;
+}
+
+
+/**************************************************************************/
+void cz230s_state::pcm_w(offs_t offset, u8 data)
+{
+	data = (BIT(offset, 0, 6) << 2) | BIT(offset, 8, 2);
+	m_pcm->write(offset >> 10, data);
+}
+
+/**************************************************************************/
+template <int Num>
+void cz230s_state::led_w(u8 data)
+{
+	for (int i = 0; i < 8; i++)
+		m_led[Num][i] = BIT(data, i);
+}
+
+/**************************************************************************/
+void cz230s_state::port_a_w(u8 data)
+{
+	m_lcdc->data_w(data & 0xf);
+	m_lcdc->std_w(BIT(data, 5));
+	m_lcdc->ce_w(BIT(data, 6));
+
+	if (BIT(data, 7) && !BIT(m_port_a, 7))
+		led_w<0>(~data & 0x3f);
+
+	m_port_a = data;
+}
+
+/**************************************************************************/
+u8 cz230s_state::keys_r()
+{
+	u8 data = 0x3f;
+
+	if (m_key_sel < m_keys.size())
+	{
+		const u16 input = m_keys[m_key_sel].read_safe(0xfff);
+		if (BIT(m_key_mux, 0))
+			data &= (input & 0x3f);
+		if (BIT(m_key_mux, 1))
+			data &= (input >> 6);
+	}
+
+	return data;
+}
+
+/**************************************************************************/
+template <int Row>
+CUSTOM_INPUT_MEMBER(cz230s_state::keys_row_r)
+{
+	u8 data = 0xff;
+
+	for (int i = 0; i < 5; i++)
+		if (BIT(m_key_sel, i))
+			data &= m_keys[i].read_safe(0xff);
+
+	return BIT(data, Row);
+}
+
+/**************************************************************************/
+template <int Row>
+u8 cz230s_state::keys_analog_r()
+{
+	return keys_row_r<Row>() ? 0xff : 0x00;
+}
+
+/**************************************************************************/
+INPUT_CHANGED_MEMBER(cz230s_state::rhythm_w)
+{
+	if (!oldval && newval)
+		m_rhythm = param;
+}
+
+/**************************************************************************/
+INPUT_CHANGED_MEMBER(cz230s_state::mode_w)
+{
+	if (!oldval && newval)
+		m_mode = param;
+}
+
+/**************************************************************************/
+void cz230s_state::cassette_w(int state)
+{
+	m_cassette->output(state ? -1.0 : 1.0);
+}
+
+/**************************************************************************/
+void cz230s_state::cassette_motor_w(int state)
+{
+	m_cassette->change_state(state ? CASSETTE_MOTOR_ENABLED : CASSETTE_MOTOR_DISABLED, CASSETTE_MASK_MOTOR);
+}
+
+/**************************************************************************/
+CUSTOM_INPUT_MEMBER(cz230s_state::cassette_r)
+{
+	return m_cassette->input() > 0 ? 0 : 1;
+}
+
+/**************************************************************************/
+void cz230s_state::render_w(int state)
+{
+	if(!state)
+		return;
+
+	for (int digit = 0; digit < 6; digit++)
```

Comment:
Another minor formatting thing – this `if` statement is missing the space before the opening parenthesis of its control expression.